File: src/python/sascalc.py
'''
    SASSIE: Copyright (C) 2011 Joseph E. Curtis, Ph.D. 

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
'''
import sys
import numpy
import sasop
import sasmath

#	SASCALC
#
#	12/10/2009	--	initial coding			:	jc
#
#	 1         2         3         4         5         6         7
#LC4567890123456789012345678901234567890123456789012345678901234567890123456789
#								       *      **
'''
	Sascalc contains the classes and methods to calculate various
	atomic and molecular properties and perform basic translation
	and rotation operations on instances of objects described in
	the sasmol module


'''

class Prop(object):

	'''
	This class calculates various properties of the object.

	calcmass calculates the mass of the object.
	
	calccom calculates the com of the object

	calcrg calculates the radius of gyration

	calcpmi calculates the principal moments of inertia (pmi)

	calcminmax calculates the minimum and maximum coordinates of an object

    calcformula determines the molecular formula based on elements

	'''	

	def calcmass(self):

		'''
		http://physics.nist.gov/cgi-bin/Compositions/stand_alone.pl?ele=&ascii=html&isotype=some

		standard atomic weight is based on the natural istopic composition

		NOTE: deuterium is 'D' 2H1 and '1H' is 1H1, all other elements 
	have their natural abundance weight. These elements are located
		at the end of the dictionary.

		'''
		
		standard_atomic_weight = self.amu() 
		self._totalmass=0.0
		self._mass=numpy.zeros(len(self._element),numpy.float)
		count=0	
		for element in self._element:
			if element in standard_atomic_weight:
				self._totalmass=self._totalmass+standard_atomic_weight[element]	
				self._mass[count]=standard_atomic_weight[element]
				count=count+1
			else:
				count=count+1

		return self._totalmass

	def calccom(self,frame):
	
		'''	
		This method calculates the center of mass of the object.  
		
		'''

		if(self._totalmass <=0.0):
			self.calcmass()
		x=self._coor[frame,:,0] ; y=self._coor[frame,:,1] ; z=self._coor[frame,:,2]	
		comx=numpy.sum(self._mass*x)/self._totalmass	
		comy=numpy.sum(self._mass*y)/self._totalmass
		comz=numpy.sum(self._mass*z)/self._totalmass
		self._com=numpy.array([comx,comy,comz],numpy.float)
	
		del x,y,z,comx,comy,comz
		
		return self._com

	def calcrg(self,frame):
		'''	
		This method calculates the radius of gyration of the object
		
		'''
		self._com = self.calccom(frame)
	
		if(self._natoms>0):
			brg2=((self._coor[frame,:,:]-self._com)*(self._coor[frame,:,:]-self._com))
			self._rg=numpy.sqrt(numpy.sum(numpy.sum(brg2))/self._natoms)
		
		return self._rg

	def calcrmsd(self,other):
		'''	
		This method calculates the radius root mean square deviation
		of one set of coordinates compared to another

		self contains the coordinates of set 1
		other contains the coordinates of set 2
	
		the number of coordinates in each set must be equal	
	
		To use this over multiple frames you must call this function
		repeatedly.
		
		'''
	###	OPEN	Add frame support here?
		try:
			dxyz=((self._coor-other._coor)*(self._coor-other._coor))
			self._rmsd=numpy.sqrt((numpy.sum(dxyz))/self._natoms)
		except:
			if(self._natoms!=other._natoms):
				print 'number of atoms in (1) != (2)'
				print 'rmsd not calculated: None returned'
				print 'number of atoms in self is < 1'
				print 'number of atoms in other is < 1'
				self._rmsd=None

		return self._rmsd

	def calcpmi(self,frame):
		'''	
		This method calculates the principal moments of inertia
		of the object. It uses the center method from sasop.Move
		to center the molecule. 

		The present method is designated for the non-linear system with
		non-singular moment of inertia matrix only. For the linear systems, it
		will return eigenvectors and I as None.  Testing for non-None return
		values should be done in the calling method.
		
		'''

		com = self.calccom(frame)

		sasop.Move.center(self,frame)

        	Ixx=0.0 ; Iyy=0.0 ; Izz=0.0
        	Ixy=0.0 ; Ixz=0.0 ; Iyz=0.0
        	Iyx=0.0 ; Izx=0.0 ; Izy=0.0
	
        	for i in range(self._natoms):
                	xp=self._coor[frame,i,0] ; yp=self._coor[frame,i,1] ; zp=self._coor[frame,i,2]
                	Ixx=Ixx+self._mass[i]*(yp*yp+zp*zp)
                	Iyy=Iyy+self._mass[i]*(xp*xp+zp*zp)
                	Izz=Izz+self._mass[i]*(xp*xp+yp*yp)
	
                	Ixy=Ixy-self._mass[i]*xp*yp
                	Ixz=Ixz-self._mass[i]*xp*zp
                	Iyz=Iyz-self._mass[i]*yp*zp
                	Iyx=Iyx-self._mass[i]*yp*xp
                	Izx=Izx-self._mass[i]*zp*xp
                	Izy=Izy-self._mass[i]*zp*yp
	
		I=numpy.array([[Ixx,Ixy,Ixz],[Iyx,Iyy,Iyz],[Izx,Izy,Izz]])
	   

		if numpy.linalg.matrix_rank(I)<3:
			print("You are requesting the pmi calculation for a singular system.")
			print("The eigen-decomposition of this system is not defined")

			uk = None ; ak = None ; I = None 
		else:
			uk,ak=numpy.linalg.eig(I)

		sasop.Move.moveto(self,frame,com)

		return uk,ak,I

#uk =  [  1.30834716e+07   1.91993314e+08   1.85015201e+08]
#ak =  [[-0.08711655 -0.97104917  0.22242802]
# [-0.512547    0.23514759  0.82583363]
# [ 0.85422847  0.04206103  0.51819358]]
#I =  [[  1.90290278e+08  -9.27036143e+06   1.25097100e+07]
# [ -9.27036143e+06   1.40233826e+08   7.53462714e+07]
# [  1.25097100e+07   7.53462714e+07   5.95678835e+07]]

	def calcminmax(self):
		'''	
		This method calculates the min and max of of the object in (x,y,z)
		
		A numpy array of min and max values are returned
		
		'''

	###	TEST	Need a test case for this ... seems to work	

		min_x=numpy.min(self._coor[:,:,0]); max_x=numpy.max(self._coor[:,:,0])		
		min_y=numpy.min(self._coor[:,:,1]); max_y=numpy.max(self._coor[:,:,1])		
		min_z=numpy.min(self._coor[:,:,2]); max_z=numpy.max(self._coor[:,:,2])		

		self._total_minimum=numpy.array([min_x,min_y,min_z])
		self._total_maximum=numpy.array([max_x,max_y,max_z])

# for min3.pdb:
# 	min =  [-24.50785265 -69.52322867 -93.88458512]
#	max =  [ 26.43814735  54.09377133  91.46441488]
#	

		return [self._total_minimum,self._total_maximum]


	def calcminmax_frame(self,frame):
	
		'''	
		This method calculates the min and max of frame=frame of the object in (x,y,z)
		
		A numpy array of min and max values are returned
		
		'''

	###	TEST	Need a test case for this ... seems to work	

		min_x=numpy.min(self._coor[frame,:,0]); max_x=numpy.max(self._coor[frame,:,0])		
		min_y=numpy.min(self._coor[frame,:,1]); max_y=numpy.max(self._coor[frame,:,1])		
		min_z=numpy.min(self._coor[frame,:,2]); max_z=numpy.max(self._coor[frame,:,2])		

		self._minimum=numpy.array([min_x,min_y,min_z])
		self._maximum=numpy.array([max_x,max_y,max_z])

# for min3.pdb:
# 	min =  [-24.50785265 -69.52322867 -93.88458512]
#	max =  [ 26.43814735  54.09377133  91.46441488]
#	

		return [self._minimum,self._maximum]

	def calc_minmax_all_steps(self,dcdfilename,**kwargs):

		if 'pdb' in kwargs:
			file_type = 'pdb'
			number_of_frames = self.number_of_frames()
		else:
			file_type = 'dcd'
			dcdfilepointer_array = self.open_dcd_read(dcdfilename)
			dcdfile = dcdfilepointer_array[0]
			number_of_frames = dcdfilepointer_array[2]

		min_x = None ; min_y = None ; min_z = None
		max_x = None ; max_y = None ; max_z = None
	
		for i in xrange(number_of_frames):

			if(file_type=='dcd'):	
				self.read_dcd_step(dcdfilepointer_array,i)
				this_minmax = self.calcminmax_frame(0)
			else:	
				this_minmax = self.calcminmax_frame(i)

			this_min_x = this_minmax[0][0] ; this_max_x = this_minmax[1][0]		
			this_min_y = this_minmax[0][1] ; this_max_y = this_minmax[1][1]
			this_min_z = this_minmax[0][2] ; this_max_z = this_minmax[1][2]
	
			if((min_x == None) or (this_min_x < min_x)): min_x = this_min_x 
			if((min_y == None) or (this_min_y < min_y)): min_y = this_min_y 
			if((min_z == None) or (this_min_z < min_z)): min_z = this_min_z 
	
			if((max_x == None) or (this_max_x > max_x)): max_x = this_max_x 
			if((max_y == None) or (this_max_y > max_y)): max_y = this_max_y 
			if((max_z == None) or (this_max_z > max_z)): max_z = this_max_z 

		if(file_type=='dcd'):	
			self.close_dcd_read(dcdfile)
		
		self._minimum = numpy.array([min_x,min_y,min_z])
		self._maximum = numpy.array([max_x,max_y,max_z])
	
		return [self._minimum,self._maximum]


	def calculate_residue_charge(self):
		'''

		Method to sum the atomic charges and assign the net charge of the
		resiude to a new variable that is attached to each atom.

		usage:
		
		single_molecule.calculate_residue_charge()
	
		residue_charge = single_molecule.residue_charge()

		print 'res-charge = ',residue_charge[0]

		'''

		resid = self.resid()
		natoms = self.natoms()

###
###	needs a gentle failure if atom_charge() does not exist
###
		atom_charge = self.atom_charge()

		charge_sum = 0.0
		charge_residue_sum = []
		last_resid = resid[0]

		for i in xrange(natoms):
			this_resid = resid[i]
			this_charge = atom_charge[i]

			if(this_resid != last_resid or i == natoms-1):
				charge_residue_sum.append([last_resid,charge_sum])
				charge_sum = this_charge
				last_resid = this_resid
			else:
				charge_sum += this_charge

		last_resid = resid[0]
		charge_residue = []

		for i in xrange(natoms):
			this_resid = resid[i]
			for j in xrange(len(charge_residue_sum)):
				if(this_resid == charge_residue_sum[j][0]):
					charge_residue.append(charge_residue_sum[j][1])
					continue

		self.setResidue_charge(numpy.array(charge_residue,numpy.float32))

		return
	
        def calcformula(self):
		'''

		Method to determine the number of each element in the molecule
                '''

                my_formula = {}

		for element in self._element:

                        if element in my_formula:
                            my_formula[element] += 1
                        else:
                            my_formula[element] = 1

                self.setFormula(my_formula)

                return my_formula

File: src/cpp/sasmol/src/sascalc.cpp
#include <sascalc.h>
#include <sasmol.h>
#include <map>
#include <Eigen/Dense>

void
initialize_amu(std::map<std::string,float> &amu)
{
        amu["H"]  = 1.007947 ;
        amu["He"] = 4.0026022 ;
        amu["Li"] = 6.9412 ; 
        amu["Be"] = 9.0121823 ; 
        amu["B"]  = 10.8117 ;
        amu["C"]  = 12.01078 ;
        amu["N"]  = 14.00672 ;
        amu["O"]  = 15.99943 ;
        amu["F"]  = 18.99840325 ;
        amu["Ne"] = 20.17976 ;
        amu["Na"] = 22.9897702 ; 
        amu["Mg"] = 24.30506 ;
        amu["Al"] = 26.9815382 ; 
        amu["Si"] = 28.08553 ;
        amu["P"]  = 30.9737612; 
        amu["S"]  = 32.0655 ; 
        amu["Cl"] = 35.4532 ; 
        amu["Ar"] = 39.9481 ;
        amu["K"]  = 39.09831 ; 
        amu["Ca"] = 40.0784 ; 
        amu["Sc"] = 44.9559108 ;
        amu["Ti"] = 47.8671 ;
        amu["V"]  = 50.94151 ; 
        amu["Cr"] = 51.99616 ; 
        amu["Mn"] = 54.9380499 ; 
        amu["Fe"] = 55.8452 ;
        amu["Co"] = 58.9332009 ; 
        amu["Ni"] = 58.69342 ; 
        amu["Cu"] = 63.5463 ; 
        amu["Zn"] = 65.4094 ;
        amu["Ga"] = 69.7231 ; 
        amu["Ge"] = 72.641 ; 
        amu["As"] = 74.921602 ; 
        amu["Se"] = 78.963 ;
        amu["Br"] = 79.9041 ; 
        amu["Kr"] = 83.7982 ; 
        amu["Rb"] = 85.46783 ;
        amu["Sr"] = 87.621 ;
        amu["Y"]  = 88.905852 ; 
        amu["Zr"] = 91.2242 ; 
        amu["Nb"] = 92.906382 ;
        amu["Mo"] = 95.942 ; 
        amu["Tc"] = 98.0 ; 
        amu["Ru"] = 101.072 ; 
        amu["Rh"] = 102.905502 ; 
        amu["Pd"] = 106.421 ; 
        amu["Ag"] = 107.86822 ; 
        amu["Cd"] = 112.4118 ; 
        amu["In"] = 114.8183 ; 
        amu["Sn"] = 118.7107 ;
        amu["Sb"] = 121.7601 ; 
        amu["Te"] = 127.603 ; 
        amu["I"]  = 126.904473 ; 
        amu["Xe"] = 131.2936 ;
        amu["Cs"] = 132.905452 ; 
        amu["Ba"] = 137.3277 ; 
        amu["La"] = 138.90552 ; 
        amu["Ce"] = 140.1161 ; 
        amu["Pr"] = 140.907652 ; 
        amu["Nd"] = 144.243 ; 
        amu["Pm"] = 145.0 ; 
        amu["Sm"] = 150.363 ; 
        amu["Eu"] = 151.9641 ; 
        amu["Gd"] = 157.253 ; 
        amu["Tb"] = 158.925342 ; 
        amu["Dy"] = 162.5001 ;
        amu["Ho"] = 164.930322 ; 
        amu["Er"] = 167.2593 ; 
        amu["Tm"] = 168.93421 ; 
        amu["Yb"] = 173.043 ; 
        amu["Lu"] = 174.9671 ; 
        amu["Hf"] = 174.9671 ; 
        amu["Ta"] = 180.94791 ; 
        amu["W"]  = 183.841 ; 
        amu["Re"] = 186.2071 ; 
        amu["Os"] = 190.233 ; 
        amu["Ir"] = 192.2173 ; 
        amu["Pt"] = 195.0782 ; 
        amu["Au"] = 196.966552 ; 
        amu["Hg"] = 200.592 ; 
        amu["Tl"] = 204.38332 ; 
        amu["Pb"] = 207.21 ; 
        amu["Bi"] = 208.980382 ; 
        amu["Po"] = 209.0 ; 
        amu["At"] = 210.0 ; 
        amu["Rn"] = 222.0 ; 
        amu["Fr"] = 223.0 ; 
        amu["Ra"] = 226.0 ; 
        amu["Ac"] = 227.0 ; 
        amu["Th"] = 232.03811 ; 
        amu["Pa"] = 231.035882 ; 
        amu["U"]  = 238.028913 ; 
        amu["D"]  = 2.01410177804 ; 
        amu["1H"] = 1.00782503214 ;
} ;

///
/// @par Detailed description 
/// ... 
/// @param [in, out] (param1) ...
/// @return ...
/// @note ...
std::vector<float>
sascalc::Prop::
calc_com(int &frame)
{

    if(_total_mass() <= 0.0) {
        calc_mass() ;
        std::cout << "com tot mass = " << _total_mass() << std::endl ;
    }

    float comx = 0.0, comy = 0.0, comz = 0.0 ;

    try
    {
        comx = (_atom_mass()*_x().col(frame)).sum() / _total_mass() ;    
        comy = (_atom_mass()*_y().col(frame)).sum() / _total_mass() ;    
        comz = (_atom_mass()*_z().col(frame)).sum() / _total_mass() ;    
    }
    catch(std::overflow_error e) {
        std::cout << "failed in calc_com" << std::endl ;
        std::cout << "total_mass = " << _total_mass() << std::endl ;
        std::cout << e.what() << " -> ";
        std::vector<float> com = {comx, comy, comz} ;
        return com ;
    }

    std::vector<float> com = {comx, comy, comz} ;

    return com ;
}


///
/// @par Detailed description 
/// ... 
/// @param [in, out] (param1) ...
/// @return ...
/// @note ...
float
sascalc::Prop::
calc_rg(int &frame){

    std::vector<float> com = calc_com(frame) ;

    float rgx = ((_x().col(frame) - com[0]).pow(2.0)).sum() ;
    float rgy = ((_y().col(frame) - com[1]).pow(2.0)).sum() ;
    float rgz = ((_z().col(frame) - com[2]).pow(2.0)).sum() ;
    
    float rg = sqrt((rgx + rgy + rgz) / _natoms()) ;    

    return rg ;
}


///
/// @par Detailed description 
/// ... 
/// @param [in, out] (param1) ...
/// @return ...
/// @note ...
void
sascalc::Prop::
calc_mass()
{
    /*
        http://physics.nist.gov/cgi-bin/Compositions/stand_alone.pl?ele=&ascii=html&isotype=some

        standard atomic weight is based on the natural istopic composition

        NOTE: deuterium is 'D' 2H1 and '1H' is 1H1, all other elements 
        have their natural abundance weight. These elements are located
        at the end of the dictionary.

    */    
        
    std::map<std::string,float> standard_atomic_weight ;
    initialize_amu(standard_atomic_weight) ;

    float total_mass = 0.0 ;
    std::string this_element ;
    float this_mass ;

    _atom_mass().setZero(_natoms(), 1);

    for(int i = 0 ; i < _natoms() ; ++i)
    {
        this_element = _atom_selement()[i] ;
        this_mass = standard_atomic_weight.find(this_element)->second ;
        total_mass += this_mass ;
        _atom_mass()[i] = this_mass ;
    }

    _total_mass() = total_mass ;
    
    return ;
}

///
/// @par Detailed description 
/// ... 
/// @param [in, out] (param1) ...
/// @return ...
/// @note ...
void
sascalc::Prop::
calc_pmi(int &frame)
{
    std::vector<float> com = calc_com(frame) ;

    _x().col(frame) -= com[0] ;
    _y().col(frame) -= com[1] ;
    _z().col(frame) -= com[2] ;

    float Ixx = (_atom_mass()*(_y().col(frame)*_y().col(frame) + _z().col(frame)*_z().col(frame))).sum() ;
    float Iyy = (_atom_mass()*(_x().col(frame)*_x().col(frame) + _z().col(frame)*_z().col(frame))).sum() ;
    float Izz = (_atom_mass()*(_x().col(frame)*_x().col(frame) + _y().col(frame)*_y().col(frame))).sum() ;

    float Ixy = (-_atom_mass()*(_x().col(frame)*_y().col(frame))).sum() ;
    float Ixz = (-_atom_mass()*(_x().col(frame)*_z().col(frame))).sum() ;
    float Iyz = (-_atom_mass()*(_y().col(frame)*_z().col(frame))).sum() ;
    float Iyx = (-_atom_mass()*(_y().col(frame)*_x().col(frame))).sum() ;
    float Izx = (-_atom_mass()*(_z().col(frame)*_x().col(frame))).sum() ;
    float Izy = (-_atom_mass()*(_z().col(frame)*_y().col(frame))).sum() ;
    
    Eigen::Matrix3f I ;

    I << Ixx, Ixy, Ixz,
         Iyx, Iyy, Iyz,
         Izx, Izy, Izz;

    Eigen::SelfAdjointEigenSolver<Eigen::Matrix3f> eigensolver(I);

    if (eigensolver.info() != Eigen::Success) 
    {
        std::cout << "PMI calculation failed" << std::endl ;
        return ;
    }

    _uk() = eigensolver.eigenvalues() ; // .col(0) ;
    _ak() = eigensolver.eigenvectors() ;

//    std::cout << "The eigenvalues of I are:\n" << eigensolver.eigenvalues() << std::endl;
//    std::cout << "Here's a matrix whose columns are eigenvectors of I \n"
//    << "corresponding to these eigenvalues:\n"
//    << eigensolver.eigenvectors() << std::endl;

    _x().col(frame) += com[0] ;
    _y().col(frame) += com[1] ;
    _z().col(frame) += com[2] ;

    return ;
}

///
/// @par Detailed description 
/// ... 
/// @param [in, out] (param1) ...
/// @return ...
/// @note ...
int
sascalc::Prop::
self_overlap(float &cutoff, int &frame){

    int check = 0 ;     // overlapped == 1
                  // default == no overlap will return == 0 

    float dist ; 
    float this_x , that_x ;
    float this_y , that_y ;
    float this_z , that_z ;
    float dx2, dy2, dz2 ;
    int count = 1 ;

    for(int i = 0 ; i != _natoms()-1 ; ++i)
    {
        this_x = _x()(i,frame) ;
        this_y = _y()(i,frame) ;
        this_z = _z()(i,frame) ;

        for(int j = i+1 ; j != _natoms() ; ++j)
        {
            that_x = _x()(j,frame) ;
            that_y = _y()(j,frame) ;
            that_z = _z()(j,frame) ;

            dx2 = (this_x - that_x)*(this_x - that_x) ;
            dy2 = (this_y - that_y)*(this_y - that_y) ;
            dz2 = (this_z - that_z)*(this_z - that_z) ;

            dist = sqrt(dx2+dy2+dz2) ;

            if(dist < cutoff) 
            {
                std::cout << "count = " << count << std::endl ;
                std::cout << "this_atom = " << _atom_name()[count-1] << std::endl ;
                std::cout << "that_atom = " << _atom_name()[count] << std::endl ;
                std::cout << "dist = " << dist << std::endl ;
                std::cout << "this_x = " << this_x << std::endl ;
                std::cout << "this_y = " << this_y << std::endl ;
                std::cout << "this_z = " << this_z << std::endl ;
                std::cout << "that_x = " << that_x << std::endl ;
                std::cout << "that_y = " << that_y << std::endl ;
                std::cout << "that_z = " << that_z << std::endl ;
                check = 1 ; 
                return check ;
            }
        }    
//        std::cout << count << " " ;
        count++ ;

    }
    std::cout << std::endl ;
    
    return check ;

}

///
/// @par Detailed description 
/// ... 
/// @param [in, out] (param1) ...
/// @return ...
/// @note ...
float
sascalc::Prop::
calc_rmsd(sasmol::SasMol &mol, int &frame){

    float dx = (_x().col(frame)-mol._x().col(frame)).pow(2.0).sum() ;
    float dy = (_y().col(frame)-mol._y().col(frame)).pow(2.0).sum() ;
    float dz = (_z().col(frame)-mol._z().col(frame)).pow(2.0).sum() ;

    float rmsd = sqrt((dx + dy + dz)/_natoms()) ;

    return rmsd ;
}

///
/// @par Detailed description 
/// ... 
/// @param [in, out] (param1) ...
/// @return ...
/// @note ...
std::vector<std::vector<float >>
sascalc::Prop::
calc_minmax(int &frame){

    float min_x = _x().col(frame).minCoeff() ; float max_x = _x().col(frame).maxCoeff() ;
    float min_y = _y().col(frame).minCoeff() ; float max_y = _y().col(frame).maxCoeff() ;
    float min_z = _z().col(frame).minCoeff() ; float max_z = _z().col(frame).maxCoeff() ;

    std::vector<float> min_vector ;
    std::vector<float> max_vector ;

    min_vector.push_back(min_x) ; min_vector.push_back(min_y) ; min_vector.push_back(min_z) ;
    max_vector.push_back(max_x) ; max_vector.push_back(max_y) ; max_vector.push_back(max_z) ;

    std::vector<std::vector<float >> minmax ;
    minmax.push_back(min_vector) ; minmax.push_back(max_vector) ;

    return minmax ;
}

/*
to do:
    def calc_minmax_all_steps() --> apbs.py & sld.py
    def calculate_residue_charge() --> mmc only
*/

